plugins/Compressor/Compressor.cpp

```diff
@@ -0,0 +1,630 @@
+	"Lost Robot <r94231@gmail.com>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	NULL,
+	NULL
+} ;
+
+}
+
+
+CompressorEffect::CompressorEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key) :
+	Effect(&compressor_plugin_descriptor, parent, key),
+	m_compressorControls(this)
+{
+	m_sampleRate = Engine::mixer()->processingSampleRate();
+
+	m_yL[0] = m_yL[1] = COMP_NOISE_FLOOR;
+
+	// These will be resized later
+	m_rms[0] = new RmsHelper(1);
+	m_rms[1] = new RmsHelper(1);
+
+	// 200 ms
+	m_crestTimeConst = exp(-1.f / (0.2f * m_sampleRate));
+
+	connect(&m_compressorControls.m_attackModel, SIGNAL(dataChanged()), this, SLOT(calcAttack()));
+	connect(&m_compressorControls.m_releaseModel, SIGNAL(dataChanged()), this, SLOT(calcRelease()));
+	connect(&m_compressorControls.m_holdModel, SIGNAL(dataChanged()), this, SLOT(calcHold()));
+	connect(&m_compressorControls.m_ratioModel, SIGNAL(dataChanged()), this, SLOT(calcRatio()));
+	connect(&m_compressorControls.m_rangeModel, SIGNAL(dataChanged()), this, SLOT(calcRange()));
+	connect(&m_compressorControls.m_rmsModel, SIGNAL(dataChanged()), this, SLOT(resizeRMS()));
+	connect(&m_compressorControls.m_lookaheadLengthModel, SIGNAL(dataChanged()), this, SLOT(calcLookaheadLength()));
+	connect(&m_compressorControls.m_thresholdModel, SIGNAL(dataChanged()), this, SLOT(calcThreshold()));
+	connect(&m_compressorControls.m_kneeModel, SIGNAL(dataChanged()), this, SLOT(calcKnee()));
+	connect(&m_compressorControls.m_outGainModel, SIGNAL(dataChanged()), this, SLOT(calcOutGain()));
+	connect(&m_compressorControls.m_inGainModel, SIGNAL(dataChanged()), this, SLOT(calcInGain()));
+	connect(&m_compressorControls.m_tiltModel, SIGNAL(dataChanged()), this, SLOT(calcTiltCoeffs()));
+	connect(&m_compressorControls.m_tiltFreqModel, SIGNAL(dataChanged()), this, SLOT(calcTiltCoeffs()));
+	connect(&m_compressorControls.m_limiterModel, SIGNAL(dataChanged()), this, SLOT(redrawKnee()));
+	connect(&m_compressorControls.m_mixModel, SIGNAL(dataChanged()), this, SLOT(calcMix()));
+
+	connect(&m_compressorControls.m_autoAttackModel, SIGNAL(dataChanged()), this, SLOT(calcAutoAttack()));
+	connect(&m_compressorControls.m_autoReleaseModel, SIGNAL(dataChanged()), this, SLOT(calcAutoRelease()));
+
+	connect(&m_compressorControls.m_thresholdModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_ratioModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_kneeModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+
+	connect(Engine::mixer(), SIGNAL(sampleRateChanged()), this, SLOT(changeSampleRate()));
+	emit changeSampleRate();
+}
+
+
+
+
+CompressorEffect::~CompressorEffect()
+{
+	delete m_rms[0];
+	delete m_rms[1];
+}
+
+
+float CompressorEffect::timeToCoeff(float time)
+{
+	// Convert time in milliseconds to applicable lowpass coefficient
+	return expf(COMP_LOG / (time * m_sampleRate * 0.001f));
```

Comment:
You may precalculate `COMP_LOG / (m_sampleRate * 0.001f)` for optimization.
Also, it's hard to know that the unit of `time` without looking at the implementation. You may add `ms` in the function name or the parameter name.